IM_calculation/scripts/A_KonnoMatricesComputation.py

```diff
@@ -0,0 +1,178 @@
+    # Check if the center frequency is exactly part of the provided
+    # frequencies. This will result in a division by 0. The limit of f->f_c is
+    # one.
+    smoothing_window[frequencies == center_frequency] = 1.0
+    # Also a frequency of zero will result in a logarithm of -inf. The limit of
+    # f->0 with f_c!=0 is zero.
+    smoothing_window[frequencies == 0.0] = 0.0
+    # Normalize to one if wished.
+    if normalize:
+        smoothing_window /= smoothing_window.sum()
+    return smoothing_window
+
+
+def calculate_smoothing_matrix(
+    frequencies, bandwidth=40.0, normalize=False, fn: str = None
+):
+    """
+    Calculates a len(frequencies) x len(frequencies) matrix with the Konno &
+    Ohmachi window for each frequency as the center frequency.
+
+    Any spectrum with the same frequency bins as this matrix can later be
+    smoothed by using
+    :func:`~obspy.signal.konnoohmachismoothing.apply_smoothing_matrix`.
+
+    This also works for many spectra stored in one large matrix and is even
+    more efficient.
+
+    This makes it very efficient for smoothing the same spectra again and again
+    but it comes with a high memory consumption for larger frequency arrays!
+
+    :type frequencies: :class:`numpy.ndarray` (float32 or float64)
+    :param frequencies:
+        The input frequencies.
+    :type bandwidth: float
+    :param bandwidth:
+        Determines the width of the smoothing peak. Lower values result in a
+        broader peak. Must be greater than 0. Defaults to 40.
+    :type normalize: bool, optional
+    :param normalize:
+        The Konno-Ohmachi smoothing window is normalized on a logarithmic
+        scale. Set this parameter to True to normalize it on a normal scale.
+        Default to False.
+    :param fn: The name of the file to save the matrix to. Cached in ram otherwise. Useful for systems with low ram
+    """
+    # Create matrix to be filled with smoothing entries.
+
+    sm_matrix = np.empty((len(frequencies), len(frequencies)), frequencies.dtype)
+    if fn:
+        np.save(fn, sm_matrix)
+        sm_matrix = np.load(fn, mmap_mode="r+")
+        # sm_matrix = np.memmap(fn, dtype=frequencies.dtype, mode='w+', shape=(len(frequencies), len(frequencies)))
+        print("Generated")
+
+    for _i, freq in enumerate(frequencies):
+        sm_matrix[_i, :] = konno_ohmachi_smoothing_window(
+            frequencies, freq, bandwidth, normalize=normalize
+        )
+        # print(sm_matrix[_i])
+    return sm_matrix
+
+
+def getSmoothMatrix(ft_freq, fn: str = None):
+    # creates Konno Ohmachi matrix
+    smooth_matrix = calculate_smoothing_matrix(
+        ft_freq, bandwidth=20, normalize=True, fn=fn
+    )
+    return smooth_matrix
+
+
+def createKonnoMatrix_single(ft_len, fn: str = None):
+    # creates several Konno Ohmachi matrices
+    dt = 0.005
+    ft_freq = np.arange(0, ft_len / 2 + 1) * (1.0 / (ft_len * dt))
```

Comment:
Change this line to:
    ft_freq = np.fft.rfftfreq(ft_len, dt)

This is more generic and works when ft_len is both even and odd. Current implementation only works for even. It should usually be even but this is better anyway.